pkg/registry/generic/registry/store_test.go

```diff
@@ -546,6 +546,230 @@ func TestStoreDelete(t *testing.T) {
+		Spec:       api.PodSpec{NodeName: "machine"},
+	}
+	_, _, err = registry.Update(testContext, updatedPodWithFinalizer)
+	if err != nil {
+		t.Errorf("Unexpected error: %v", err)
+	}
+
+	// the object should still exist, because it still has a finalizer
+	obj, err = registry.Get(testContext, podWithFinalizer.Name)
+	if err != nil {
+		t.Errorf("Unexpected error: %v", err)
+	}
+	podWithFinalizer, ok = obj.(*api.Pod)
+	if !ok {
+		t.Errorf("Unexpected object: %#v", obj)
+	}
+
+	podWithNoFinalizer := &api.Pod{
+		ObjectMeta: api.ObjectMeta{Name: "foo", ResourceVersion: podWithFinalizer.ObjectMeta.ResourceVersion},
+		Spec:       api.PodSpec{NodeName: "anothermachine"},
+	}
+	_, _, err = registry.Update(testContext, podWithNoFinalizer)
+	if err != nil {
+		t.Errorf("Unexpected error: %v", err)
+	}
+	// the pod should be removed, because it's finalizer is removed
+	_, err = registry.Get(testContext, podWithFinalizer.Name)
+	if !errors.IsNotFound(err) {
+		t.Errorf("Unexpected error: %v", err)
+	}
+}
+
+func TestStoreDeleteWithOrphanDependents(t *testing.T) {
+	podWithOrphanFinalizer := func(name string) *api.Pod {
+		return &api.Pod{
+			ObjectMeta: api.ObjectMeta{Name: name, Finalizers: []string{"foo.com/x", api.OrphanFinalizer, "bar.com/y"}},
+			Spec:       api.PodSpec{NodeName: "machine"},
+		}
+	}
+	podWithOtherFinalizers := func(name string) *api.Pod {
+		return &api.Pod{
+			ObjectMeta: api.ObjectMeta{Name: name, Finalizers: []string{"foo.com/x", "bar.com/y"}},
+			Spec:       api.PodSpec{NodeName: "machine"},
+		}
+	}
+	podWithNoFinalizer := func(name string) *api.Pod {
+		return &api.Pod{
+			ObjectMeta: api.ObjectMeta{Name: name},
+			Spec:       api.PodSpec{NodeName: "machine"},
+		}
+	}
+	podWithOnlyOrphanFinalizer := func(name string) *api.Pod {
+		return &api.Pod{
+			ObjectMeta: api.ObjectMeta{Name: name, Finalizers: []string{api.OrphanFinalizer}},
+			Spec:       api.PodSpec{NodeName: "machine"},
+		}
+	}
+	trueVar, falseVar := true, false
+	orphanOptions := &api.DeleteOptions{OrphanDependents: &trueVar}
+	nonOrphanOptions := &api.DeleteOptions{OrphanDependents: &falseVar}
+
+	testcases := []struct {
+		pod               *api.Pod
+		options           *api.DeleteOptions
+		expectNotFound    bool
+		updatedFinalizers []string
+	}{
+		// cases run with DeleteOptions.OrphanDedependents=true
+		{
+			podWithOrphanFinalizer("pod1"),
+			orphanOptions,
+			false,
+			[]string{"foo.com/x", api.OrphanFinalizer, "bar.com/y"},
+		},
+		{
+			podWithOtherFinalizers("pod2"),
+			orphanOptions,
+			false,
+			[]string{"foo.com/x", "bar.com/y", api.OrphanFinalizer},
+		},
+		{
+			podWithNoFinalizer("pod3"),
+			orphanOptions,
+			false,
+			[]string{api.OrphanFinalizer},
+		},
+		{
+			podWithOnlyOrphanFinalizer("pod4"),
+			orphanOptions,
+			false,
+			[]string{api.OrphanFinalizer},
+		},
+		// cases run with DeleteOptions.OrphanDedependents=false
```

Comment:
If it's present and set to false, maybe you should actually remove the finalizer if it's present?